+Base/QRconv.m

```diff
@@ -0,0 +1,384 @@
+     
+    if isempty(V) || all(isnan(V(:)))
+        M = [[NaN NaN]; [NaN NaN]];
+        b = [NaN; NaN];
+        
+        options_guess.Vcen = b;
+        options_guess.M = M;
+        options_guess.b = b;
+        options_guess.M2 = M;
+        options_guess.b2 = b;
+        options_guess.outliers = false(1, size(V, 2));
+        options_fit = options_guess;
+        return
+    end
+    
+    options_guess.img_W = size(img,1);
+    options_guess.img_H = size(img,2);
+    
+    % Step 5: fit a coordinate system to the positions of our QR codes.
+    [M, b, M2, b2, outliers] = majorityVoteCoordinateFit(v, V, options_guess);
+    
+    if ~any(isnan([M(:); b(:)]))
+        Vcen = invaffine([options_guess.img_H; options_guess.img_W]/2, M, b);
+    else
+        Vcen = [NaN; NaN];
+    end
+    
+    xaxis = affine([1;0], M, [0; 0]);
+    if xaxis(1) == 0
+        ang1 = (pi/2) + pi * (xaxis(2) < 0);
+    else
+        ang1 = atan(xaxis(2)/xaxis(1));
+    end
+    
+    call = options_guess.d / norm(xaxis);
+    
+    options_fit = struct('ang', ang1, 'calibration', call, 'Vcen', Vcen, 'M', M, 'b', b, 'M2', M2, 'b2', b2, 'outliers', outliers);
+end
+
+function [conv, convH, convV] = doConv(img, ang0, r, l)
+    r = ceil(r);
+
+    ang = mod(ang0, pi);
+
+    invx = sin(ang0) > 0;
+    invy = cos(ang0) > 0;
+
+    ca = cos(ang);
+    sa = sin(ang);
+
+    lx = ceil(l*ca);
+    ly = ceil(l*sa);
+
+    X = -r:lx+r;
+    Y = -r:ly+r;
+
+    % Replace with meshgrid?
+    XX = repmat(X, [length(Y) 1]);
+    YY = repmat(Y', [1 length(X)]);
+
+    B = .5;
+    sp = 2;
+    sn = 1;
+
+    % This filter descirbes the shape of a QR arm, with large holes on each end and small holes on the arm. Negative values are to increase reliablility.
+    fil =  B*circleFunc(XX, YY, 0,       0,      r) ...
+        - sn*circleFunc(XX, YY, 1*lx/8,  1*ly/8, r/3) ...
+        + sp*circleFunc(XX, YY, lx/4,    ly/4,   r/3) ...
+        - sn*circleFunc(XX, YY, 3*lx/8,  3*ly/8, r/3) ...
+        + sp*circleFunc(XX, YY, lx/2,    ly/2,   r/3) ...
+        - sn*circleFunc(XX, YY, 5*lx/8,  5*ly/8, r/3) ...
+        + sp*circleFunc(XX, YY, 3*lx/4,  3*ly/4, r/3) ...
+        - sn*circleFunc(XX, YY, 7*lx/8,  7*ly/8, r/3) ...
+        + B*circleFunc(XX, YY, lx,      ly,     r);
+
+    % Normalize
+    S1 = size(fil);
+    fil = fil - sum(sum(fil))/S1(1)/S1(2);
+    fil = fil / sqrt(sum(sum(fil.^2)));
+
+    % Uncomment this line to get a better idea of what the filter looks like.
+%     imwrite(.5 + fil/max(max(fil))/2, 'fil.png');
+
+    lx = ceil(l*ca);
+    ly = ceil(l*sa);
+
+    % The heart of the function, doing the convolution on the images with our filter, on both arms (H&V) of the QRs.
+    convH = conv2(img, fil);
+    convV = conv2(img, rot90(fil));
+
+    S = size(img);
+    X = (1:S(1)) + r;
+    Y = (1:S(2)) + r;
+
+    convH = convH(X + invx*ly, Y + invy*lx);
+    convV = convV(X + (~invy)*lx, Y + invx*ly);
+
+    conv = convH.*convH.*convH + convV.*convV.*convV;
```

Comment:
What's going on here? Might be obvious to others but just for my own edification

Reply:
Added comment.